conn.go

```diff
@@ -741,14 +756,19 @@ func (c *Conn) Handshake() Alert {
 		c.hState = state
 		logf(logTypeHandshake, "state is now %s", c.GetHsState())
 		_, connected = state.(StateConnected)
+		if connected {
+			c.state = state.(StateConnected)
+			c.handshakeComplete = true
+		}
 
-		if c.config.NonBlocking && alert == AlertStatelessRetry {
-			return AlertStatelessRetry
+		if c.config.NonBlocking {
```

Comment:
I'm a little unclear on how this affects upstream callers of `Handshake()`.  It seems like `conn.Read()`, `Listener.Accept()`, and `DialWithDialer()` all assume that `Handshake()` returning with `AlertNoAlert` means that the handshake is complete.

Those methods probably aren't suitable for use with non-blocking mode anyway (?) so maybe we just need to define a new error code that if non-blocking mode accidentally gets set, these methods will fail, but a non-blocking caller can recognize that everything's OK.

Reply:
Good point. Although QUIC doesn't use those, at least some of them would be useful even in non-blocking mode (e.g. you want to be able to `conn.Read()` after completing the handshake).

I think the best solution here is:
1. document that `Read` should only be called after the handshake is complete in non-blocking mode
2. not run `Handshake()`from `Read`, but 
```go
for {
  if alert := c.Handshake(); alert != AlertNoAlert {
    return alert
  }
  if _, ok := c.state.(StateConnected); ok {
     break
   }
}
```
This way, if `Read` is called before the handshake completes, at least it will fail at some point (with an `AlertWouldBlock`), but won't try to perform `Read` on an unconnected connection.

I'll add a commit to make things clearer.

Reply:
I just realized my idea doesn't really make sense for the listen and dial calls. Just returning an error, as you suggested, is probably the better solution. Just pushed the change, so PHAL.